python/device/device.py

```diff
@@ -0,0 +1,314 @@
+# Copyright (c) Microsoft. All rights reserved.
+# Licensed under the MIT license. See LICENSE file in the project root for
+# full license information.
+import logging
+import os
+import time
+import uuid
+import json
+import datetime
+import threading
+from concurrent.futures import ThreadPoolExecutor
+from measurement import MeasureLatency
+import dps
+import longhaul
+import reaper
+from azure.iot.device import Message
+
+logger = logging.getLogger("thief.{}".format(__name__))
+
+logging.basicConfig(level=logging.WARNING)
+# logging.getLogger("paho").setLevel(level=logging.DEBUG)
+logging.getLogger("thief").setLevel(level=logging.INFO)
+
+
+provisioning_host = os.environ["THIEF_DEVICE_PROVISIONING_HOST"]
+id_scope = os.environ["THIEF_DEVICE_ID_SCOPE"]
+group_symmetric_key = os.environ["THIEF_DEVICE_GROUP_SYMMETRIC_KEY"]
+registration_id = os.environ["THIEF_DEVICE_ID"]
+
+
+class DeviceRunMetrics(longhaul.RunMetrics):
+    """
+    Object we use internally to keep track of how a the entire test is performing.
+    """
+
+    def __init__(self):
+        super(DeviceRunMetrics, self).__init__()
+        self.d2c = longhaul.OperationMetrics()
+
+
+class DeviceRunConfig(longhaul.RunConfig):
+    """
+    Object we use internally to keep track of how the entire test is configured.
+    """
+
+    def __init__(self):
+        super(DeviceRunConfig, self).__init__()
+        self.system_telemetry_send_interval_in_seconds = 0
+        self.d2c = longhaul.OperationConfig()
+
+
+def make_new_d2c_payload(message_id):
+    """
+    Helper function to create a unique payload which can be sent up as d2c message
+    """
+    msg = {"messageId": message_id, "thiefPingback": True}
+    return Message(json.dumps(msg))
+
+
+def set_config(config):
+    """
+    Helper function which sets our configuration.  Right now, this is hardcoded.
+    Later, this will come from desired properties.
+    """
+    config.system_telemetry_send_interval_in_seconds = 10
+    config.max_run_duration_in_seconds = 0
+    config.d2c.operations_per_second = 3
+    config.d2c.timeout_interval_in_seconds = 60
+    config.d2c.failures_allowed = 0
+
+
+class DeviceApp(longhaul.LonghaulMixin, reaper.ReaperMixin):
+    """
+    Main application object
+    """
+
+    def __init__(self):
+        self.executor = ThreadPoolExecutor(max_workers=128)
+        self.done = threading.Event()
+        self.client = None
+        self.metrics = DeviceRunMetrics()
+        self.config = DeviceRunConfig()
+        self.d2c_set_lock = threading.Lock()
+        self.d2c_confirmed = set()
+        self.d2c_unconfirmed = set()
+        self.last_heartbeat = time.time()
+        self.shutdown_event = threading.Event()
+
+        self.start_reaper()
+
+    def update_initial_properties(self):
+        """
+        Update reported properties at the start of a run
+        """
+        # toto: update these values
```

Comment:
is this still todo?

Reply:
No it is a [toto](https://www.youtube.com/watch?v=FTQbiNvZqaY), which is totally different!

Reply:
Yes, it's still a toto.  That dog is going to come through one of these days.  It's also a todo

---
In reply to: [491170812](https://github.com/Azure/iot-sdk-longhaul/pull/9#discussion_r491170812) [](ancestors = 491170812)